docs/api/v3.rst

```diff
@@ -0,0 +1,558 @@
+API v3
+======
+
+The Read the Docs API uses :abbr:`REST (Representational State Transfer)`.
+JSON is returned by all API responses including errors
+and HTTP response status codes are to designate success and failure.
+
+
+Authentication and authorization
+--------------------------------
+
+Requests to the Read the Docs public API are for public and private information
+and do require authentication even for requesting public information.
```

Comment:
I'm -0 on this. I think having a public API that isn't authed fits a lot of use cases nicely, and I haven't heard an argument for why we should exclude it in this design. I'm fine to heavily rate limit anon access so it isn't built into processes, but I think for learning etc. having it open is good.

Reply:
Also, is it possible to use this API on a documentation page as designed? eg. I'm a user who wants to get the version listing of my project in my own JS, can I do that?

Reply:
To me, not requiring authentication duplicates a mistake we made with API v1 and v2. Now that we are deprecating API v1 and v2, there's no way to know who is making outdated API calls because the calls aren't authenticated. Other than broadcast messages in our docs and on the blog, we can't warn users specifically that their API calls are going to break after a certain date.

Secondly, I question the usefulness of an API that allows browsing all projects, versions, and builds rather than "my" projects, versions, and builds. For example, our current v2 endpoints would be a lot more useful if they returned the few projects where you are a maintainer (at least by default) rather than `/api/v2/projects/` returning all projects and requiring you to page through them or filter by a single slug.

Still, maybe the right approach is to *heavily* rate limit the API when used anonymously such that almost everyone will get a token.

> Also, is it possible to use this API on a documentation page as designed? eg. I'm a user who wants to get the version listing of my project in my own JS, can I do that?

With authentication, probably not. However, how does one know which projects are "my" projects without authentication?

Reply:
> With authentication, probably not. However, how does one know which projects are "my" projects without authentication?

`GET /api/v3/projects/<slug>versions/`

Reply:
That gets **a project** but there's no concept of **my project** without auth.

Reply:
well sure, but presumably the JS embedded in a project's docs know what the project is :)

Reply:
I agree in general auth is useful, but I think we need to support unauth'd endpoints for a lot of pretty important use cases, or have some kind of publishable token.

Reply:
Wait a second... I think Eric has a point. How are we going to use our own API (for the footer for example) if we do not allow anonymous access? Will we allow anonymous access _only_ on the endpoints that we know that are going to be used for any doc page?

I'm not sure if we should raise this topic again or not(*), but I tend to agree with Eric that it's good to have the API open. If you are designing a theme for Read the Docs, you won't be able to query anything via the API. Let's say that you want to create your custom version menu flyout:

    /api/v3/projects/pip/versions/?active=true

won't be available. What we would do in these cases?


(*) is all about communication with users?

Reply:
There are some endpoints that need to be unauthed -- footer api, ads api, etc. But I'm :-1: on a public endpoints like `/api/v3/projects` listing. @davidfischer sums up the problems nicely.

I think requiring auth is pretty accepted API design at this point from a user perspective, and I don't feel that a public API provides enough value that we need to complicate our design for this use case. There is however a lot of value for us to keep things simple. If there is anything in particular that requires a public API, we should enumerate these though.

> What we would do in these cases?

Use a API token :) Part of the intersphinx token work was to eventually pass in a generated token to the build process for this particular case, using our API from the build process.